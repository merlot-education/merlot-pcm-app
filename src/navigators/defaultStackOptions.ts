import { StackNavigationOptions } from '@react-navigation/stack'
import { ColorPallet } from '../theme/theme'

const defaultStackOptions: StackNavigationOptions = {
  headerStyle: {
    backgroundColor: ColorPallet.brand.primary,
    elevation: 0,
    shadowOpacity: 0,
    borderBottomWidth: 0,
  },
  headerTintColor: ColorPallet.grayscale.white,
  headerBackTitleVisible: false,
  headerTitleAlign: 'center',
}

export default defaultStackOptions
