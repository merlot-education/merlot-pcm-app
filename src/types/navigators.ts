export type TabStackParams = {
  HomeTab: undefined
  ContactsTab: undefined
  ScanTab: undefined
  CredentialsTab: undefined
  SettingsTab: undefined
}

export type AuthenticateStackParams = {
  'Enter Pin': { setAuthenticated: (auth: boolean) => void }
}

export type ContactStackParams = {
  Contacts: undefined
  'Contact Details': { connectionId: string }
}

export type CredentialStackParams = {
  Credentials: undefined
  'Credential Details': { credentialId: string }
}

export type HomeStackParams = {
  Home: undefined
  Notifications: undefined
  'Credential Offer': { credentialId: string }
  'Proof Request': { proofId: string }
}

export type ScanStackParams = {
  Scan: undefined
}

export type SettingsStackParams = {
  Settings: undefined
  Language: undefined
}
