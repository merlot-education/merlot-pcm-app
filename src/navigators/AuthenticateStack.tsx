import React from 'react';
import { createStackNavigator } from '@react-navigation/stack';
import { AuthenticateStackParams, Screens } from '../types/navigators';
import PinEnter from '../screens/PinEnter';

import defaultStackOptions from './defaultStackOptions';

interface AuthenticateStackProps {
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
}

const Stack = createStackNavigator<AuthenticateStackParams>();

const AuthenticateStack: React.FC<AuthenticateStackProps> = ({
  setAuthenticated,
}) => {
  return (
    <Stack.Navigator
      screenOptions={{
        ...defaultStackOptions,
        presentation: 'transparentModal',
        headerShown: false,
      }}
    >
      <Stack.Screen
        name={Screens.EnterPin}
        component={PinEnter}
        initialParams={{ setAuthenticated }}
      />
    </Stack.Navigator>
  );
};

export default AuthenticateStack;
