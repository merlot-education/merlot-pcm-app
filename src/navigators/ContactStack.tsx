import { createStackNavigator } from '@react-navigation/stack'
import React from 'react'

import ListContacts from '../screens/ListContacts'
import { ContactStackParams, Screens } from '../types/navigators'
import ContactDetails from '../screens/ContactDetails'

import defaultStackOptions from './defaultStackOptions'
import { ColorPallet } from '../theme/theme'

const Stack = createStackNavigator<ContactStackParams>()

const ContactStack: React.FC = () => {
  return (
    <Stack.Navigator screenOptions={{ ...defaultStackOptions }}>
      <Stack.Screen
        name={Screens.ListContacts}
        component={ListContacts}
        options={() => ({
          title: 'Connection',
          headerTintColor: ColorPallet.baseColors.white,
          headerShown: true,
          headerLeft: () => false,
          rightLeft: () => false,
        })}
      />
      <Stack.Screen name={Screens.ContactDetails} component={ContactDetails} />
    </Stack.Navigator>
  )
}

export default ContactStack
