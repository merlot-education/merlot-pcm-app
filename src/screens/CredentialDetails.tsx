import type { StackScreenProps } from '@react-navigation/stack'

import { useCredentialById } from '@aries-framework/react-hooks'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import Toast from 'react-native-toast-message'
import Record from '../components/record/Record'
import { ToastType } from '../components/toast/BaseToast'
import CredentialCard from '../components/misc/CredentialCard'
import { ColorPallet, TextTheme } from '../theme/theme'
import { CredentialStackParams, Screens } from '../types/navigators'

type CredentialDetailsProps = StackScreenProps<
  CredentialStackParams,
  Screens.CredentialDetails
>

const styles = StyleSheet.create({
  headerText: {
    ...TextTheme.normal,
  },
  footerText: {
    ...TextTheme.normal,
    paddingTop: 16,
  },
  linkContainer: {
    minHeight: TextTheme.normal.fontSize,
    paddingVertical: 2,
  },
  link: {
    ...TextTheme.normal,
    color: ColorPallet.brand.link,
  },
})

const CredentialDetails: React.FC<CredentialDetailsProps> = ({
  navigation,
  route,
}) => {
  const { t } = useTranslation()
  const { credentialId } = route.params
  const credential = useCredentialById(credentialId)

  //   const getCredentialRecord = (
  //     credentialId?: string,
  //   ): CredentialRecord | void => {
  //     try {
  //       if (!credentialId) {
  //         throw new Error(t('CredentialOffer.CredentialNotFound'))
  //       }

  //       return useCredentialById(credentialId)
  //     } catch (e: unknown) {
  //       Toast.show({
  //         type: ToastType.Error,
  //         text1: t('Global.Failure'),
  //         text2: (e as Error)?.message || t('CredentialOffer.CredentialNotFound'),
  //       })

  //       navigation.goBack()
  //     }
  //   }

  if (!route.params.credentialId) {
    Toast.show({
      type: ToastType.Error,
      text1: t('Global.Failure'),
      text2: t('CredentialOffer.CredentialNotFound'),
    })

    navigation.goBack()
    return null
  }

  //   const credential = useCredentialById(route.params.credentialId)

  if (!credential) {
    Toast.show({
      type: ToastType.Error,
      text1: t('Global.Failure'),
      text2: t('CredentialOffer.CredentialNotFound'),
    })

    navigation.goBack()
    return null
  }

  return (
    <Record
      header={() => (
        <CredentialCard
          credential={credential}
          style={{ marginHorizontal: 15, marginTop: 16 }}
        />
      )}
      footer={() => (
        <View style={{ marginBottom: 30 }}>
          {/* <TouchableOpacity activeOpacity={1}>
            <Text
              style={[
                styles.footerText,
                styles.link,
                { color: ColorPallet.semantic.error },
              ]}
            >
              {t('CredentialDetails.RemoveFromWallet')}
            </Text>
          </TouchableOpacity> */}
        </View>
      )}
      attributes={credential.credentialAttributes}
      hideAttributeValues
    />
  )
}

export default CredentialDetails
