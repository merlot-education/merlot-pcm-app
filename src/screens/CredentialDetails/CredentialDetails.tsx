import type { StackScreenProps } from '@react-navigation/stack';

import { useAgent, useCredentialById } from '@aries-framework/react-hooks';
import React, { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { View, Text, StyleSheet, Platform, ScrollView } from 'react-native';
import CredentialCard from '../../components/misc/CredentialCard';
import { CredentialStackParams, Screens } from '../../types/navigators';
import { warningToast, errorToast } from '../../utils/toast';
import { ColorPallet, TextTheme } from '../../theme/theme';
import Accordion from '../../components/accordion/Accordion';
import { credentialDefinition } from '../../utils/helpers';
import { RecordHistory } from '../../types/record';

type CredentialDetailsProps = StackScreenProps<
  CredentialStackParams,
  Screens.CredentialDetails
>;

const CredentialDetails: React.FC<CredentialDetailsProps> = ({
  navigation,
  route,
}) => {
  const { t } = useTranslation();
  const { credentialId } = route.params;
  const credential = useCredentialById(credentialId);
  const { agent } = useAgent();
  const [history, setHistory] = useState<RecordHistory[]>([]);

  const getCredentialHistory = useCallback(async () => {
    try {
      const data = await agent?.genericRecords.findAllByQuery({
        credentialRecordId: credential?.credentials[0].credentialRecordId,
      });

      const history: any[] = [];

      // Add all credential records to the history
      data?.forEach(record => history.push(...record.content.records));

      // Filter out the credential records which have property credentialLabel
      const filteredHistory = history.filter(record => record?.credentialLabel);

      // Sort history by timestamp
      const sortedHistory = filteredHistory.sort(
        (x, y) =>
          new Date(y.timestamp).valueOf() - new Date(x.timestamp).valueOf(),
      );

      setHistory(sortedHistory);
    } catch (error) {
      errorToast(t('credential.get.error'));
    }
  }, [agent, credential, t]);

  useEffect(() => {
    getCredentialHistory();
  }, [getCredentialHistory]);

  if (!route.params.credentialId) {
    warningToast(t('CredentialOffer.CredentialNotFound'));
    navigation.goBack();
    return null;
  }
  if (!credential) {
    errorToast(t('CredentialOffer.CredentialNotFound'));
    navigation.goBack();
    return null;
  }

  return (
    <ScrollView contentContainerStyle={styles.scrollView}>
      <CredentialCard
        credential={credential}
        style={styles.credentialCardView}
      />
      <View style={Platform.OS === 'android' ? styles.card : styles.cardIos}>
        <Accordion title="Info" innerAccordion={false}>
          <View>
            <View style={styles.innerContainer}>
              <Text style={styles.attribute}>Credential </Text>
              <Text style={styles.attribute}>
                {credentialDefinition(credential)?.split(':')[4]}
              </Text>
            </View>
            <View style={styles.divider} />
            {credential?.credentialAttributes?.map(item => {
              return (
                <View style={styles.innerContainer}>
                  <Text style={styles.attribute}>{item.name}</Text>
                  <Text style={styles.attribute}>{item.value}</Text>
                </View>
              );
            })}
          </View>
        </Accordion>
      </View>
      <View style={Platform.OS === 'android' ? styles.card : styles.cardIos}>
        <Accordion title="Activities" innerAccordion={false}>
          {history.map((item, index) => {
            return (
              <View>
                <Accordion
                  title={item.connectionLabel}
                  date={new Date(item.timestamp)}
                  status={item.status}
                  innerAccordion>
                  {Object.entries(item.attributes).map(([key, value]) => {
                    return (
                      <View style={styles.innerContainer}>
                        <Text style={styles.attribute}>{key}</Text>
                        <Text style={styles.attribute}>{value.toString()}</Text>
                      </View>
                    );
                  })}
                </Accordion>
                {history.length - 1 !== index && (
                  <View style={styles.divider} />
                )}
              </View>
            );
          })}
        </Accordion>
      </View>
    </ScrollView>
  );
};

export default CredentialDetails;

const styles = StyleSheet.create({
  card: {
    backgroundColor: ColorPallet.baseColors.white,
    borderRadius: 10,
    elevation: 3,
    padding: 10,
    marginVertical: 10,
    width: '90%',
    alignSelf: 'center',
  },
  cardIos: {
    backgroundColor: ColorPallet.baseColors.white,
    shadowColor: ColorPallet.baseColors.black,
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    borderRadius: 10,
    padding: 10,
    marginVertical: 10,
    width: '90%',
    alignSelf: 'center',
  },
  text: {
    fontSize: 18,
    marginBottom: 20,
  },
  safeArea: {
    flex: 1,
  },
  heading: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 10,
  },
  hidden: {
    height: 0,
  },
  list: {
    overflow: 'hidden',
  },
  sectionTitle: {
    ...TextTheme.normal,
    fontWeight: 'bold',
    color: ColorPallet.baseColors.black,
    marginLeft: '5%',
  },
  sectionSubTitle: {
    ...TextTheme.caption,
    color: ColorPallet.baseColors.black,
    marginLeft: '5%',
  },
  sectionDescription: {
    ...TextTheme.caption,
    color: ColorPallet.baseColors.black,
    height: 30,
    marginLeft: '5%',
  },
  divider: {
    borderBottomColor: ColorPallet.baseColors.lightGrey,
    borderBottomWidth: 1,
    width: '100%',
  },
  credentialCardView: {
    marginHorizontal: 15,
    marginTop: 16,
  },
  innerContainer: {
    flexDirection: 'row',
    marginVertical: 5,
  },
  attribute: {
    width: '50%',
    color: ColorPallet.baseColors.black,
  },
  scrollView: {
    paddingBottom: 30,
  },
});
