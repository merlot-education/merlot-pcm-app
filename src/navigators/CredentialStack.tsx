import { createStackNavigator } from '@react-navigation/stack';
import React from 'react';
import CredentialDetails from '../screens/CredentialDetails';
import ListCredentials from '../screens/ListCredentials';
import { CredentialStackParams, Screens } from '../types/navigators';

import defaultStackOptions from './defaultStackOptions';

const Stack = createStackNavigator<CredentialStackParams>();

const CredentialStack: React.FC = () => {
  return (
    <Stack.Navigator screenOptions={{ ...defaultStackOptions }}>
      <Stack.Screen name={Screens.Credentials} component={ListCredentials} />
      <Stack.Screen
        name={Screens.CredentialDetails}
        component={CredentialDetails}
      />
    </Stack.Navigator>
  );
};

export default CredentialStack;
