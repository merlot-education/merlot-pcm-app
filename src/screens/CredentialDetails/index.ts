import CredentialDetails from './CredentialDetails'

export default CredentialDetails
