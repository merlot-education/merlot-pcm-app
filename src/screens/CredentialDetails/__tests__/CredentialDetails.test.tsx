import React from 'react'
import {
  CredentialMetadataKeys,
  CredentialExchangeRecord,
  CredentialState,
  CredentialPreviewAttribute,
} from '@aries-framework/core'
import { useNavigation } from '@react-navigation/core'
import { create } from 'react-test-renderer'
import CredentialDetails from '../CredentialDetails'
import { Screens } from '../../../types/navigators'

const credentialRecord = new CredentialExchangeRecord({
  connectionId: '28790bfe-1345-4c64-b21a-7d98982b3894',
  threadId: 'threadId',
  state: CredentialState.Done,
  credentialAttributes: [
    new CredentialPreviewAttribute({
      name: 'age',
      value: '25',
    }),
  ],
  protocolVersion: 'v1',
})

credentialRecord.metadata.set(CredentialMetadataKeys.IndyCredential, {
  credentialDefinitionId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TA',
  schemaId: 'TL1EaPFCZ8Si5aUrqScBDt:2:testschema:1.0',
})

jest.mock('i18next', () => ({
  use: jest.fn(() => ({})),
  init: jest.fn(() => ({})),
  t: k => k,
}))

jest.mock('react-native-toast-message', () => ({
  show: jest.fn(),
  hide: jest.fn(),
}))
describe('CredentialDetails', () => {
  jest.mock('react-native', () => {
    const RN = jest.requireActual('react-native')

    return Object.setPrototypeOf(
      {
        Alert: {
          ...RN.Alert,
          alert: jest.fn(),
        },
      },
      RN,
    )
  })
  it('credential details render correctly', () => {
    const tree = create(
      <CredentialDetails
        route={{
          params: { credentialId: credentialRecord.id },
          key: '',
          name: Screens.CredentialDetails,
        }}
        navigation={useNavigation()}
      />,
    )

    expect(tree).toMatchSnapshot()
  })
})
