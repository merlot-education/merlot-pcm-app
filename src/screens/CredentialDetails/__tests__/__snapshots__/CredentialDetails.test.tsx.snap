// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialDetails credential details render correctly 1`] = `null`;
